File: samples/java/ZoneWorld/Server/src/main/java/systems/zlink/samples/zoneworld/server/ops/OpsSession.java
package systems.zlink.samples.zoneworld.server.ops;

import systems.zlink.framework.channels.ZLinkFanoutClient;
import systems.zlink.framework.messaging.ZLinkMessage;
import systems.zlink.framework.streams.ZLinkSession;
import systems.zlink.framework.streams.ZLinkSessionContext;
import systems.zlink.framework.streams.ZLinkSessionDispatchContext;
import systems.zlink.framework.streams.ZLinkStreamError;
import systems.zlink.samples.zoneworld.server.configuration.MaintenanceStore;
import systems.zlink.samples.zoneworld.server.configuration.NodeRegistry;
import systems.zlink.samples.zoneworld.shared.Messages;
import systems.zlink.samples.zoneworld.shared.ZoneWorldNames;

import java.util.List;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public final class OpsSession implements ZLinkSession {
    private final ZLinkSessionContext context;
    private final NodeRegistry registry;
    private final MaintenanceStore maintenance;
    private final ZLinkFanoutClient fanout;
    private final OpsConsoleRegistry consoles;

    public OpsSession(
            ZLinkSessionContext context,
            NodeRegistry registry,
            MaintenanceStore maintenance,
            ZLinkFanoutClient fanout,
            OpsConsoleRegistry consoles) {
        this.context = context;
        this.registry = registry;
        this.maintenance = maintenance;
        this.fanout = fanout;
        this.consoles = consoles;
    }

    @Override
    public ZLinkSessionContext context() {
        return context;
    }

    @Override
    public CompletionStage<Void> onConnected() {
        consoles.add(context);
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onDisconnected() {
        consoles.remove(context);
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onError(ZLinkStreamError error) {
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onDispatch(
            ZLinkSessionDispatchContext dispatch, ZLinkMessage payload) {
        return switch (dispatch.packetName()) {
            case "WatchNodesReq" -> watch();
            case "AnnounceWorldReq" -> announce(payload.decode(Messages.AnnounceWorldReq.class));
            case "SetMaintenanceReq" ->
                    setMaintenance(payload.decode(Messages.SetMaintenanceReq.class));
            case "NodeDiagnosticsReq" ->
                    diagnostics(payload.decode(Messages.NodeDiagnosticsReq.class));
            default ->
                    throw new IllegalStateException(
                            "Unknown ZoneWorld Ops packet: " + dispatch.packetName());
        };
    }

    private CompletionStage<Void> watch() {
        List<Messages.NodeView> nodes = registry.snapshot();
        return reply(new Messages.WatchNodesRes(nodes))
                .thenCompose(ignored -> consoles.replay(context, nodes));
    }

    private CompletionStage<Void> announce(Messages.AnnounceWorldReq request) {
        String id = UUID.randomUUID().toString();
        return fanout.publish(
                        ZoneWorldNames.BROADCAST_CHANNEL,
                        ZoneWorldNames.ANNOUNCE_TOPIC,
                        new Messages.WorldAnnounceEvent(id, request.text()))
                .submit()
                .thenCompose(ignored -> reply(new Messages.AnnounceWorldRes(id)));
    }

    private CompletionStage<Void> setMaintenance(Messages.SetMaintenanceReq request) {
        Messages.NodeView node = registry.find(request.nodeId());
        if (node == null) {
            return reply(
                    new Messages.SetMaintenanceRes(
                            request.nodeId(), false, List.of(), "UnknownNode"));
        }
        // The desired state is committed to the store; what the console shows for the node
        // stays whatever the node itself last reported, so an observed maintenance value is
        // always the node's own and never the operator's intent echoed back.
        maintenance.set(request.nodeId(), request.enabled());
        // --8<-- [start:doc-zw-ops-publish]
        return fanout.publish(
                        ZoneWorldNames.BROADCAST_CHANNEL,
                        ZoneWorldNames.MAINTENANCE_TOPIC,
                        new Messages.NodeMaintenanceChangedEvent(
                                request.nodeId(), request.enabled()))
                .submit()
                .thenCompose(
                        ignored ->
                                reply(
                                        new Messages.SetMaintenanceRes(
                                                request.nodeId(),
                                                request.enabled(),
                                                node.zones(),
                                                null)));
        // --8<-- [end:doc-zw-ops-publish]
    }

    private CompletionStage<Void> diagnostics(Messages.NodeDiagnosticsReq request) {
        Messages.NodeView node = registry.find(request.nodeId());
        if (node == null) {
            return reply(
                    new Messages.NodeDiagnosticsRes(
                            request.nodeId(),
                            List.of(),
                            0,
                            maintenance.get(request.nodeId()),
                            "UnknownNode"));
        }
        return reply(
                new Messages.NodeDiagnosticsRes(
                        node.nodeId(), node.zones(), node.playerCount(), node.maintenance(), null));
    }

    private CompletionStage<Void> reply(Object message) {
        return context.client().reply(message).submit();
    }
}

File: samples/kotlin/ZoneWorld/Server/src/main/kotlin/systems/zlink/samples/kotlin/zoneworld/server/ops/OpsSession.kt
package systems.zlink.samples.kotlin.zoneworld.server.ops

import java.util.UUID
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionStage
import java.util.concurrent.CopyOnWriteArrayList
import systems.zlink.framework.channels.ZLinkFanoutClient
import systems.zlink.framework.messaging.ZLinkMessage
import systems.zlink.framework.streams.ZLinkSession
import systems.zlink.framework.streams.ZLinkSessionContext
import systems.zlink.framework.streams.ZLinkSessionDispatchContext
import systems.zlink.framework.streams.ZLinkStreamError
import systems.zlink.samples.kotlin.zoneworld.server.configuration.MaintenanceStore
import systems.zlink.samples.kotlin.zoneworld.server.configuration.NodeRegistry
import systems.zlink.samples.kotlin.zoneworld.shared.Messages
import systems.zlink.samples.kotlin.zoneworld.shared.ZoneWorldNames

class OpsSession(
    private val sessionContext: ZLinkSessionContext,
    private val registry: NodeRegistry,
    private val maintenance: MaintenanceStore,
    private val fanout: ZLinkFanoutClient,
    private val consoles: OpsConsoleRegistry,
) : ZLinkSession {
    override fun context() = sessionContext

    override fun onConnected(): CompletionStage<Void> {
        consoles.add(sessionContext)
        return CompletableFuture.completedFuture(null)
    }

    override fun onDisconnected(): CompletionStage<Void> {
        consoles.remove(sessionContext)
        return CompletableFuture.completedFuture(null)
    }

    override fun onError(error: ZLinkStreamError): CompletionStage<Void> =
        CompletableFuture.completedFuture<Void>(null)

    override fun onDispatch(
        dispatch: ZLinkSessionDispatchContext,
        payload: ZLinkMessage,
    ): CompletionStage<Void> =
        when (dispatch.packetName()) {
            "WatchNodesReq" -> watch()
            "AnnounceWorldReq" -> announce(payload.decode(Messages.AnnounceWorldReq::class.java))
            "SetMaintenanceReq" ->
                setMaintenance(payload.decode(Messages.SetMaintenanceReq::class.java))
            "NodeDiagnosticsReq" ->
                diagnostics(payload.decode(Messages.NodeDiagnosticsReq::class.java))
            else -> error("unknown ZoneWorld Ops packet ${dispatch.packetName()}")
        }

    private fun watch(): CompletionStage<Void> {
        val nodes = registry.snapshot()
        return reply(Messages.WatchNodesRes(nodes)).thenCompose {
            consoles.replay(sessionContext, nodes)
        }
    }

    private fun announce(request: Messages.AnnounceWorldReq): CompletionStage<Void> {
        val id = UUID.randomUUID().toString()
        return fanout
            .publish(
                ZoneWorldNames.BROADCAST_CHANNEL,
                ZoneWorldNames.ANNOUNCE_TOPIC,
                Messages.WorldAnnounceEvent(id, request.text),
            )
            .submit()
            .thenCompose { reply(Messages.AnnounceWorldRes(id)) }
    }

    private fun setMaintenance(request: Messages.SetMaintenanceReq): CompletionStage<Void> {
        val node =
            registry.find(request.nodeId)
                ?: return reply(
                    Messages.SetMaintenanceRes(request.nodeId, false, emptyList(), "UnknownNode")
                )
        // The desired state is committed to the store; what the console shows for the node
        // stays whatever the node itself last reported, so an observed maintenance value is
        // always the node's own and never the operator's intent echoed back.
        maintenance.set(request.nodeId, request.enabled)
        // --8<-- [start:doc-zw-ops-publish]
        return fanout
            .publish(
                ZoneWorldNames.BROADCAST_CHANNEL,
                ZoneWorldNames.MAINTENANCE_TOPIC,
                Messages.NodeMaintenanceChangedEvent(request.nodeId, request.enabled),
            )
            .submit()
            .thenCompose {
                reply(Messages.SetMaintenanceRes(request.nodeId, request.enabled, node.zones))
            }
        // --8<-- [end:doc-zw-ops-publish]
    }

    private fun diagnostics(request: Messages.NodeDiagnosticsReq): CompletionStage<Void> {
        val node = registry.find(request.nodeId)
        return if (node == null)
            reply(
                Messages.NodeDiagnosticsRes(
                    request.nodeId,
                    emptyList(),
                    0,
                    maintenance.get(request.nodeId),
                    "UnknownNode",
                )
            )
        else
            reply(
                Messages.NodeDiagnosticsRes(
                    node.nodeId,
                    node.zones,
                    node.playerCount,
                    node.maintenance,
                )
            )
    }

    private fun reply(message: Any): CompletionStage<Void> =
        sessionContext.client().reply(message).submit()
}

class OpsConsoleRegistry {
    private val sessions = CopyOnWriteArrayList<ZLinkSessionContext>()
    private val alerts = CopyOnWriteArrayList<Messages.NodeAlertNotify>()

    fun add(context: ZLinkSessionContext) {
        sessions += context
    }

    fun remove(context: ZLinkSessionContext) {
        sessions -= context
    }

    fun record(alert: Messages.NodeAlertNotify) {
        alerts += alert
    }

    fun broadcast(message: Any) {
        sessions.toList().forEach { session ->
            try {
                session.client().send(message).submit().exceptionally { null }
            } catch (_: RuntimeException) {
                // A stale console cannot prevent later live consoles from receiving
                // this best-effort status or alert notification.
            }
        }
    }

    fun replay(
        context: ZLinkSessionContext,
        nodes: List<Messages.NodeView>,
    ): CompletionStage<Void> {
        var sends: CompletionStage<Void> = CompletableFuture.completedFuture(null)
        nodes.toList().forEach { node ->
            sends =
                sends.thenCompose {
                    context
                        .client()
                        .send(
                            Messages.NodeStatusNotify(
                                node.nodeId,
                                node.registered,
                                node.connected,
                                node.maintenance,
                                node.zones,
                                node.playerCount,
                            )
                        )
                        .submit()
                }
        }
        alerts.toList().forEach { alert ->
            sends = sends.thenCompose { context.client().send(alert).submit() }
        }
        return sends
    }
}
